REGRESSION (iOS 17): Dragging a calendar event from Spotlight into Mail doesn’t attach an .ics file

whsieh opened the pull request:
#### c63a695fb5e5f85bb7dc75efc3bcb008696e5703
<pre>
REGRESSION (iOS 17): Dragging a calendar event from Spotlight into Mail doesn’t attach an .ics file
<a href="https://bugs.webkit.org/show_bug.cgi?id=272308">https://bugs.webkit.org/show_bug.cgi?id=272308</a>
<a href="https://rdar.apple.com/120436702">rdar://120436702</a>

Reviewed by Abrar Protyasha.

In iOS 17, Spotlight no longer sets `-[NSItemProvider suggestedName]` when dragging calendar events.
This causes a drop from Spotlight into Mail compose to fall back to inserting plain text instead of
a file, since:

1. The dropped item has a plain text representation.
2. The dropped item does not have a preferred attachment representation type specified.
3. The dropped item does not have a suggested file name specified.

The fact that Spotlight has stopped vending this suggested name in iOS 17 is being tracked in
<a href="https://rdar.apple.com/126050147">rdar://126050147</a>; in the meantime, we can also improve this experience in WebKit by treating dropped
calendar invites as one of the types that defaults to being inserted as an attachment, rather than
text (similar to PDF and VCard files).

* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::shouldTreatAsAttachmentByDefault):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:
(TestWebKitAPI::calendarInviteForTesting):
(TestWebKitAPI::TEST(WKAttachmentTestsIOS, InsertDroppedCalendarInviteAsAttachment)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/event.ics: Added.

Add an API test to verify the fix.

Canonical link: <a href="https://commits.webkit.org/277187@main">https://commits.webkit.org/277187@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/870c4ee2da039aad144bc9efd6785bb69cbbd450

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46928 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26092 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49558 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49610 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42976 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49235 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31033 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23561 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38220 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40421 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19530 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20776 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41561 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4974 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41944 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51483 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21945 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18293 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45514 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23228 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44497 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10364 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23926 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -305,8 +305,10 @@ static void readURLAlongsideAttachmentIfNecessary(PasteboardWebContentReader& re
+    RetainPtr type = [UTType typeWithIdentifier:typeIdentifier];
+    return [type conformsToType:UTTypeVCard]
+        || [type conformsToType:UTTypePDF]
+        || [[type preferredMIMEType] isEqualToString:@"text/calendar"];
```
Would `[type conformsToType:UTTypeCalendarEvent]` not work?

aprotyas replied:
Oh, maybe that's what

> 2. The dropped item does not have a preferred attachment representation type specified.

is about

whsieh replied:
Oh, that's actually about Spotlight not explicitly setting `-[NSItemProvider preferredPresentationStyle]` to `UIPreferredPresentationStyleAttachment`.

I'll check to see if `UTTypeCalendarEvent` works…

whsieh replied:
Yep, that works — I'll check that instead, and factor the code out into a for loop like this:

```
static bool shouldTreatAsAttachmentByDefault(const String& typeIdentifier)
{
    RetainPtr type = [UTType typeWithIdentifier:typeIdentifier];
    for (UTType *attachmentType : std::array { UTTypeVCard, UTTypePDF, UTTypeCalendarEvent }) {
        if ([type conformsToType:attachmentType])
            return true;
    }
    return false;
}
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:
```diff
@@ -2590,6 +2602,22 @@ static void _generateBestRepresentationForRequest(id, SEL)
+    EXPECT_WK_STREQ("text/calendar", [webView valueOfAttribute:@"type" forQuerySelector:@"attachment"]);
+    EXPECT_EQ(1U, [simulator insertedAttachments].count);
+    _WKAttachmentInfo *info = [simulator insertedAttachments].firstObject.info;
+    EXPECT_TRUE([info.name containsString:@".ics"]);
```
Doesn't matter for the test, but I'm curious: What is the inserted file name?

whsieh replied:
Currently, an (otherwise-unnamed) calendar event gets dropped as `"ICS File.ics"`.

whsieh committed:
REGRESSION (iOS 17): Dragging a calendar event from Spotlight into Mail doesn’t attach an .ics file
https://bugs.webkit.org/show_bug.cgi?id=272308
rdar://120436702

Reviewed by Abrar Protyasha.

In iOS 17, Spotlight no longer sets `-[NSItemProvider suggestedName]` when dragging calendar events.
This causes a drop from Spotlight into Mail compose to fall back to inserting plain text instead of
a file, since:

1. The dropped item has a plain text representation.
2. The dropped item does not have a preferred attachment representation type specified.
3. The dropped item does not have a suggested file name specified.

The fact that Spotlight has stopped vending this suggested name in iOS 17 is being tracked in
rdar://126050147; in the meantime, we can also improve this experience in WebKit by treating dropped
calendar invites as one of the types that defaults to being inserted as an attachment, rather than
text (similar to PDF and VCard files).

* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::shouldTreatAsAttachmentByDefault):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:
(TestWebKitAPI::calendarInviteForTesting):
(TestWebKitAPI::TEST(WKAttachmentTestsIOS, InsertDroppedCalendarInviteAsAttachment)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/event.ics: Added.

Add an API test to verify the fix.

Canonical link: https://commits.webkit.org/277187@main
